draft-ietf-mls-protocol.md

```diff
@@ -979,68 +979,124 @@ number of leaf nodes it contains.  For a given parent node, its _left
 subtree_ is the subtree with its left child as head (respectively
 _right subtree_).
 
-All trees used in this protocol are left-balanced binary trees. A
-binary tree is _full_ (and _balanced_) if its size is a power of
-two and for any parent node in the tree, its left and right subtrees
-have the same size.
-
-A binary tree is _left-balanced_ if for every
-parent, either the parent is balanced, or the left subtree of that
-parent is the largest full subtree that could be constructed from
-the leaves present in the parent's own subtree.
-Given a list of `n` items, there is a unique left-balanced
-binary tree structure with these elements as leaves.
-
-(Note that left-balanced binary trees are the same structure that is
-used for the Merkle trees in the Certificate Transparency protocol
-{{?I-D.ietf-trans-rfc6962-bis}}.)
+Every tree used in this protocol is a perfect binary tree, that is, a complete
+balanced binary tree with `2^d` leaves all at the same depth `d`.  This
+structure is unique for a given depth `d`.
+
+There are multiple ways that an implementation might represent a ratchet tree in
+memory.  A convenient property of the complete trees used here is that they can
+be represented as an array of nodes, with node relationships computed based on
+the nodes indices in the array.  A more traditional representation based on
+linked node object may also be used.  {{array-based-trees}} and
+{{link-based-trees}} provide some details on how to implement the tree
+operations required for MLS in these representations.  MLS places no
+requirements on implementations' internal representations
+of ratchet trees.  An implementation MAY use any tree representation and
+associated algorithms, as long as they produce correct protocol messages.
+
+### Ratchet Tree Nodes
+
+Each leaf node in a ratchet tree is given an _index_ (or _leaf index_), starting
+at `0` from the left to `n-1` at the right (for a tree with `n` leaves). A tree
+with `n` leaves has `2*n - 1` nodes, including parent nodes.
+
+Each node in a ratchet tree is either _blank_ (containing no value) or it holds
+an asymmetric key pair with some associated data:
+
+* A public key (for the HPKE scheme in use, see {{ciphersuites}})
+* A private key (only within the member's direct path, see {{views}})
+* A credential (only for leaf nodes)
+* An ordered list of "unmerged" leaves (see {{views}})
+* A hash of certain information about the node's parent, as of the last time the
+  node was changed (see {{parent-hash}}).
+
+Every node, regardless of whether the node is blank or populated, has
+a corresponding _hash_ that summarizes the contents of the subtree
+below that node.  The rules for computing these hashes are described
+in {{tree-hashes}}.
+
+The _resolution_ of a node is an ordered list of non-blank nodes
+that collectively cover all non-blank descendants of the node.  The resolution
+of a node is effectively a depth-first, left-first enumeration of the nearest
```

Comment:
```suggestion
of a node is effectively a left-first enumeration of the nearest
```

I find the "depth-first" misleading, e.g. in the following tree:
```
      Y
    __|__
   /     \
  _       Z
 / \     / \
A   B   C   D
```
the resolution of Y is A, B, Z and not Z, A, B.

Thinking more about it, this high-level description of what is a "resolution" is a bit weird because it doesn't takes into account unmerged leaves.
I think a better high-level description of the "resolution" would be something like "The resolution of a node contains the smallest set of keys required to encrypt a secret to its subtree".

What do you think?

Reply:
The point of the "depth-first, left-first" language is to clarify the order in which the nodes in the resolution are listed.  As I read it, the language is correct, or maybe better as "left-first, depth first", in the sense that you visit the left child before the right child, and if you find a blank left child, you descend before going to the right child.  Following those rules, you do indeed end up at [A, B, Z] in your example -- though of course the resolution of Y should be [Y] ;)

In any case, I will handle this in a follow-on, since this text is an effect of #689, not this PR.